les/utils/nodestate.go

```diff
@@ -0,0 +1,419 @@
+}
+
+func (ns *NodeStateMachine) SaveToDb() {
+	ns.lock.Lock()
+	defer ns.lock.Unlock()
+
+	for id, node := range ns.nodes {
+		if node.dirty {
+			ns.saveNode(id, node)
+		}
+	}
+}
+
+// returns after all resulting immediate changes are processed
+func (ns *NodeStateMachine) UpdateState(id enode.ID, set, reset NodeStateBitMask, timeout time.Duration) {
+	ns.lock.Lock()
+	cb := ns.updateState(id, set, reset, timeout)
+	ns.lock.Unlock()
+	cb()
+}
+
+func (ns *NodeStateMachine) updateState(id enode.ID, set, reset NodeStateBitMask, timeout time.Duration) func() {
+	node := ns.nodes[id]
+	if node == nil {
+		node = ns.newNode()
+		ns.nodes[id] = node
+		fmt.Println("updateState", id, "newNode")
+	}
+	fmt.Println("updateState", id, "state", ns.stateToString(node.state), "set", ns.stateToString(set), "reset", ns.stateToString(reset))
+	newState := (node.state & (^reset)) | set
+	if newState == node.state {
+		return func() {}
+	}
+	oldState := node.state
+	changed := oldState ^ newState
+	node.state = newState
+	// remove timers of reset states
+	ns.removeTimeouts(node, oldState&(^newState))
+	setStates := newState & (^oldState)
+	if timeout != 0 && setStates != 0 {
+		ns.addTimeout(id, setStates, timeout)
+	}
+	if newState == 0 {
+		delete(ns.nodes, id)
+		if node.db {
+			ns.deleteNode(id)
+		}
+	} else {
+		for i, f := range node.fields {
+			if f != nil {
+				if ns.nodeFieldMasks[i]&newState == 0 {
+					node.fields[i] = nil
+				}
+			}
+		}
+		if changed&ns.saveTimeout != 0 {
+			node.dirty = true
+		}
+		if changed&ns.saveImmediately != 0 {
+			ns.saveNode(id, node)
+		}
+	}
+	return func() {
+		// call state update subscription callbacks without holding the mutex
+		for _, sub := range ns.stateSubs {
+			if changed&sub.mask != 0 {
+				sub.callback(id, oldState&sub.mask, newState&sub.mask)
+			}
+		}
+	}
+}
+
+func (ns *NodeStateMachine) initState(id enode.ID, node *nodeInfo, state NodeStateBitMask) {
+	node.state = state
+	for _, sub := range ns.stateSubs {
+		if (node.state|initState)&sub.mask != 0 {
+			sub.callback(id, initState&sub.mask, node.state&sub.mask)
+		}
+	}
+}
+
+func (ns *NodeStateMachine) AddTimeout(id enode.ID, mask NodeStateBitMask, timeout time.Duration) {
+	ns.lock.Lock()
+	defer ns.lock.Unlock()
+
+	ns.addTimeout(id, mask, timeout)
+}
+
+func (ns *NodeStateMachine) addTimeout(id enode.ID, mask NodeStateBitMask, timeout time.Duration) {
+	node := ns.nodes[id]
+	if node == nil {
+		fmt.Println("addTimeout", id, "unknown")
+		return
+	}
+	fmt.Println("addTimeout", id, "state", ns.stateToString(node.state), "mask", ns.stateToString(mask), "timeout", timeout)
+	mask &= node.state
+	ns.removeTimeouts(node, mask)
+	t := &nodeStateTimeout{
+		id:   id,
+		at:   ns.clock.Now() + mclock.AbsTime(timeout),
+		mask: mask,
+	}
+	t.timer = ns.clock.AfterFunc(timeout, func() {
+		var cb func()
+		ns.lock.Lock()
+		if t.mask != 0 {
+			fmt.Println("timeout", id, "mask", ns.stateToString(t.mask))
+			cb = ns.updateState(id, 0, t.mask, 0)
```

Comment:
Just a question: So if some state are set(and timeout is specified), after the given time duration, the state will be reset and relevant callbacks are called?